mempool/src/core_mempool/mempool.rs

```diff
@@ -287,18 +304,32 @@ impl Mempool {
     ) -> MempoolStatus {
         trace!(
             LogSchema::new(LogEntry::AddTxn)
-                .txns(TxnsLog::new_txn(txn.sender(), txn.sequence_number())),
-            committed_seq_number = db_sequence_number
+                .txns(TxnsLog::new_txn(txn.sender(), txn.replay_protector())),
+            committed_seq_number = account_sequence_number
         );
 
-        // don't accept old transactions (e.g. seq is less than account's current seq_number)
-        if txn.sequence_number() < db_sequence_number {
-            return MempoolStatus::new(MempoolStatusCode::InvalidSeqNumber).with_message(format!(
-                "transaction sequence number is {}, current sequence number is  {}",
-                txn.sequence_number(),
-                db_sequence_number,
-            ));
-        }
+        if let ReplayProtector::SequenceNumber(txn_seq_num) = txn.replay_protector() {
+            // don't accept old transactions (e.g. seq is less than account's current seq_number)
+            match &account_sequence_number {
+                AccountSequenceNumberInfo::Required(account_sequence_number) => {
+                    if txn_seq_num < *account_sequence_number {
+                        return MempoolStatus::new(MempoolStatusCode::InvalidSeqNumber)
+                            .with_message(format!(
+                                "transaction sequence number is {}, current sequence number is  {}",
+                                txn_seq_num, account_sequence_number,
+                            ));
+                    }
+                },
+                AccountSequenceNumberInfo::NotRequired => {
+                    return MempoolStatus::new(MempoolStatusCode::InvalidSeqNumber).with_message(
```

Comment:
I roughly remember we allow stateless account to send txn with sequence number 0? @igor-aptos

Reply:
Actually, this account's onchain sequence number, not the transaction's sequence number. The mempool code fetches the account's sequence number for sequence number transactions. For orderless transactions, the account sequence number is not fetched.  Instead of using an Option for AccountSequenceNumberInfo, I used an enum with `Required` and `NotRequired` variants to make it easier to understand. Not sure if it's the best approach.

Reply:
Done.